### PR TITLE
Fixing up cluster rejoin behavior

### PR DESCRIPTION
**Pull request by annymsMthd**
Fixed rejoining node

### CODE REVIEW

File: src/core/Akka.Remote/Endpoint.cs [hunk 1]
```diff
@@ -824,7 +824,17 @@ protected override void PreStart()
...
             {
-                Transport.Associate(RemoteAddress, _refuseUid).ContinueWith(x => new Handle(x.Result),
+                Transport
+                    .Associate(RemoteAddress, _refuseUid)
+                    .ContinueWith(handle =>
+                    {
+                        if (handle.IsFaulted)
+                        {
+                            var inner = handle.Exception.Flatten().InnerException;
+                            return (object)new Status.Failure(new InvalidAssociationException("Association failure", inner));
```
Comment (annymsMthd) [inline src/core/Akka.Remote/Endpoint.cs]: Not sure if this is the right status failure

File: src/core/Akka.Cluster/ClusterDaemon.cs [hunk 2]
```diff
@@ -1449,11 +1454,33 @@ public double AdjustedGossipDifferentViewProbability
...
+                }
+                else
+                {
+                    _leaderActionCounter += 1;
+                    if (_leaderActionCounter == firstNotice || _leaderActionCounter%periodicNotice == 0)
+                    {
+                        _log.Info("Leader can currently not perform its duties, reachability status: [{0}], member status: [{1}]",
+                                        _latestGossip.ReachabilityExcludingDownedObservers,
+                                        _latestGossip.Members
+                                        .Select(m => string.Format("${0} ${1} seen=${2}",
```
Comment (annymsMthd) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: Need to string.Join these here

File: src/core/Akka.Cluster/ClusterDaemon.cs [hunk 3]
```diff
@@ -1449,11 +1454,33 @@ public double AdjustedGossipDifferentViewProbability
...
                 // only run the leader actions if we are the LEADER
-                if (_latestGossip.Convergence)
+                var firstNotice = 20;
+                var periodicNotice = 60;
+                if (_latestGossip.Convergence(SelfUniqueAddress))
+                {
+                    if (_leaderActionCounter >= firstNotice)
+                        _log.Info("Leader can perform its duties again");
+                    _leaderActionCounter = 0;
+                    _log.Error("Im the boss");
```
Comment (annymsMthd) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: Remove debug log before merge

File: src/core/Akka.Remote/Endpoint.cs (hunk 1 above)
Comment (Aaronontheweb) [inline src/core/Akka.Remote/Endpoint.cs]: That should be right

**Comment by Aaronontheweb** [comment]
The ClusterDomainEventPublisherSpec is still failing all tests for this build.

**Comment by annymsMthd** [comment]
@Aaronontheweb After this build i'll rebase and squash

File: src/core/Akka.Cluster/ClusterDaemon.cs [hunk 4]
```diff
@@ -1042,14 +1043,23 @@ public void Joining(UniqueAddress node, ImmutableHashSet<string> roles)
...
+                var localMember = localMembers.FirstOrDefault(m => m.Address == node.Address);
 
-                // check by address without uid to make sure that node with same host:port is not allowed
-                // to join until previous node with that host:port has been removed from the cluster
-                var alreadyMember = localMembers.Any(m => m.Address == node.Address);
-                var isUnreachable = !_latestGossip.Overview.Reachability.IsReachable(node);
-
-                if (alreadyMember) _log.Info("Existing member [{0}] is trying to join, ignoring", node);
-                else if (isUnreachable) _log.Info("Unreachable member [{0}] is trying to join, ignoring", node);
+                if (localMember != null && localMember.UniqueAddress == node)
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: were these just areas where we weren't totally up to date with the JVM version?

Comment (annymsMthd) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: yup

File: src/core/Akka.Cluster/Gossip.cs [hunk 5]
```diff
@@ -9,6 +9,7 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
+using System.Runtime.InteropServices;
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/Gossip.cs]: I assume we don't actually need this :p

**Comment by annymsMthd** [comment]
#918  linking issue

File: src/core/Akka.Cluster/Gossip.cs (hunk 5 above)
Comment (annymsMthd) [inline src/core/Akka.Cluster/Gossip.cs]: Now where did that come from?

File: src/core/Akka.Cluster/ClusterDaemon.cs [hunk 6]
```diff
@@ -1042,14 +1043,23 @@ public void Joining(UniqueAddress node, ImmutableHashSet<string> roles)
...
+                    {
+                        Sender.Tell(new InternalClusterAction.Welcome(SelfUniqueAddress, _latestGossip));
+                    }
+                }
+                else if (localMember != null)
+                {
+                    _log.Info("New incarnation of existing member [{0}] is trying to join. " +
+                        "Existing will be removed from the cluster and then new member will be allowed to join.", node);
+                    if (localMember.Status != MemberStatus.Down && localMember.Status != MemberStatus.Leaving && localMember.Status != MemberStatus.Exiting)
+                        Downing(localMember.Address);
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: so is this where the magic happens for downing the unreachable node going forward? The "big fix" as it were?

File: src/core/Akka.Cluster/Gossip.cs [hunk 7]
```diff
@@ -71,10 +72,12 @@ public static Gossip Create(ImmutableSortedSet<Member> members)
         readonly ImmutableSortedSet<Member> _members;
         readonly GossipOverview _overview;
         readonly VectorClock _version;
+        private readonly Lazy<Reachability> _reachability;
 
         public ImmutableSortedSet<Member> Members { get { return _members; } }
         public GossipOverview Overview { get { return _overview; } }
         public VectorClock Version { get { return _version; } }
+        public Reachability ReachabilityExcludingDownedObservers { get { return _reachability.Value; } }
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/Gossip.cs]: So do we actually have a concept of a "downed observer" yet or is this a bit of a placeholder?

File: src/core/Akka.Cluster/ClusterDaemon.cs (hunk 6 above)
Comment (annymsMthd) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: This is the big fix for #918 . This allows the node to rejoin

File: src/core/Akka.Cluster/Gossip.cs (hunk 7 above)
Comment (annymsMthd) [inline src/core/Akka.Cluster/Gossip.cs]: Probably a placeholder. I copied it over from akka. Check out the lazy function to see what it does.

**Pull request merged**